chainimport/headers_import.go

```diff
@@ -0,0 +1,1403 @@
+		return f.metadata, nil
+	}
+
+	if f.reader == nil {
+		return nil, errors.New("file reader not initialized")
+	}
+
+	// Read the first 6 bytes containing metadata.
+	buf := make([]byte, HeaderMetadataSize)
+	_, err := f.reader.ReadAt(buf, 0)
+	if err != nil {
+		return &HeaderMetadata{}, fmt.Errorf("failed to read "+
+			"metadata: %w", err)
+	}
+
+	// Parse metadata.
+	metadata := HeaderMetadata{
+		BitcoinChainType: BitcoinChainType(buf[0]),
+		HeaderType:       HeaderType(buf[1]),
+		StartHeight:      binary.LittleEndian.Uint32(buf[2:6]),
+	}
+
+	// Validate header type.
+	switch metadata.HeaderType {
+	case BlockHeader, FilterHeader:
+	default:
+		metadata = HeaderMetadata{}
+		err = fmt.Errorf("invalid header type: %s", metadata.HeaderType)
+	}
+
+	return &metadata, err
+}
+
+// validateTypeMatch ensures that the generic type T matches the header type in
+// the file.
+func (f *FileHeaderImportSource[T]) validateTypeMatch(
+	metadata *HeaderMetadata) error {
+
+	var empty T
+
+	// Create the appropriate header type
+	var header ChainImportHeader
+	switch any(empty).(type) {
+	case *ChainImportBlockHeader:
+		header = &ChainImportBlockHeader{}
+	case *ChainImportFilterHeader:
+		header = &ChainImportFilterHeader{}
+	default:
+		return fmt.Errorf("unsupported generic type: %T", empty)
+	}
+
+	return header.ValidateMetadata(metadata)
+}
+
+// Iterator returns an efficient iterator for sequential header access.
+func (f *FileHeaderImportSource[T]) Iterator(start,
+	end int) HeaderIterator[T] {
+
+	return &ImportSourceHeaderIterator[T]{
+		source: f, current: start, end: end,
+	}
+}
+
+func (f *FileHeaderImportSource[T]) GetHeader(index int) (T, error) {
+	var empty T
+
+	// Calculate the absolute position for this header.
+	headerSize := f.metadata.HeaderType.Size()
+	offset := HeaderMetadataSize + (index * headerSize)
+
+	// Check if requested header is within file bounds.
+	if offset+headerSize > f.fileSize {
+		return empty, fmt.Errorf("header index %d out of bounds", index)
+	}
+
+	// Calculate the actual height from index and start height.
+	height := uint32(index) + f.metadata.StartHeight
+
+	// Read header data at the calculated offset.
+	buf := make([]byte, headerSize)
+	_, err := f.reader.ReadAt(buf, int64(offset))
+	if err != nil {
+		return empty, fmt.Errorf("failed to read header at "+
+			"index %d: %w", index, err)
+	}
+	reader := bytes.NewReader(buf)
+
+	// Create the appropriate chain import header type based on metadata.
+	var header ChainImportHeader
+	switch f.metadata.HeaderType {
+	case BlockHeader:
+		header = &ChainImportBlockHeader{
+			BlockHeader: &headerfs.BlockHeader{},
+		}
+	case FilterHeader:
+		header = &ChainImportFilterHeader{
+			FilterHeader: &headerfs.FilterHeader{},
+		}
+	default:
+		return empty, fmt.Errorf("unsupported header type: %s",
+			f.metadata.HeaderType)
+	}
+
+	if err := header.Deserialize(reader, height); err != nil {
+		return empty, err
+	}
+
+	return any(header).(T), err
```

Comment:
Same here, I think if we rename `var empty T` to `var header T`, then this can just be:

```go
	// Create the appropriate chain import header type based on metadata.
	if err := header.Deserialize(reader, height); err != nil {
		return header, err
	}

	return header, err
```

If that results in a nil pointer issue, let me know, then we might need to correctly initialize the header in each `Deserialize()` method.